File: Sample/java/src/main/java/net/nend/sample/java/fullboard/FullBoardPagerActivity.java
package net.nend.sample.java.fullboard;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.AsyncTaskLoader;
import androidx.loader.content.Loader;
import androidx.viewpager.widget.ViewPager;

import net.nend.android.NendAdFullBoard;
import net.nend.android.NendAdFullBoardLoader;
import net.nend.android.NendAdFullBoardView;
import net.nend.sample.java.R;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CountDownLatch;

public class FullBoardPagerActivity extends AppCompatActivity implements LoaderManager.LoaderCallbacks<List<NendAdFullBoard>> {

    private static final String TAG = "FullBoardAd";
    private static final int AD_COUNT = 2;

    private ViewPager mPager;
    private Handler mHandler = new Handler(Looper.getMainLooper());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_full_board_pager);
        mPager = findViewById(R.id.pager);
        LoaderManager.getInstance(this).initLoader(0, null, this);
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                // Update the ad orientation.
                Objects.requireNonNull(mPager.getAdapter()).notifyDataSetChanged();
            }
        });
    }

    @NonNull
    @Override
    public Loader<List<NendAdFullBoard>> onCreateLoader(int id, Bundle args) {
        return new AdLoader(getApplicationContext());
    }

    @Override
    public void onLoadFinished(@NonNull Loader<List<NendAdFullBoard>> loader, List<NendAdFullBoard> data) {
        List<Page> pages = new ArrayList<>();
        for (int i = 0; i < 5; i++) {
            pages.add(new ContentPage());
        }
        if (AD_COUNT == data.size()) {
            pages.add(1, new AdPage(data.get(0)));
            pages.add(3, new AdPage(data.get(1)));
        } else {
            Log.d(TAG, "Couldn't obtain two ads.");
        }
        mPager.setAdapter(new Adapter(getSupportFragmentManager(), pages));
    }

    @Override
    public void onLoaderReset(@NonNull Loader<List<NendAdFullBoard>> loader) {
    }

    interface Page {
        Fragment getFragment();
    }

    public static class ContentFragment extends Fragment {

        @SuppressLint("SetTextI18n")
        @Nullable
        @Override
        public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            TextView textView = new TextView(getActivity());
            textView.setText("Content");
            textView.setGravity(Gravity.CENTER);
            return textView;
        }
    }

    public static class AdFragment extends Fragment implements NendAdFullBoardView.FullBoardAdClickListener {

        private NendAdFullBoard mAd;

        @Override
        public void onCreate(@Nullable Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            setRetainInstance(true);
        }

        @Nullable
        @Override
        public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            if (mAd != null) {
                return NendAdFullBoardView.Builder.with(requireActivity(), mAd)
                        .adClickListener(this)
                        .build();
            } else {
                return null;
            }
        }

        void setAd(NendAdFullBoard ad) {
            mAd = ad;
        }

        @Override
        public void onClickAd(NendAdFullBoard ad) {
            Log.d(TAG, "ClickAd");
        }
    }

    private static class AdLoader extends AsyncTaskLoader<List<NendAdFullBoard>> {

        private List<NendAdFullBoard> mAds;
        private NendAdFullBoardLoader mLoader;

        AdLoader(Context context) {
            super(context);
            mLoader = new NendAdFullBoardLoader(context, 485520, "a88c0bcaa2646c4ef8b2b656fd38d6785762f2ff");
        }

        @Override
        public List<NendAdFullBoard> loadInBackground() {
            final CountDownLatch latch = new CountDownLatch(AD_COUNT);
            final List<NendAdFullBoard> ads = new ArrayList<>();
            for (int i = 0; i < AD_COUNT; i++) {
                mLoader.loadAd(new NendAdFullBoardLoader.Callback() {
                    @Override
                    public void onSuccess(NendAdFullBoard ad) {
                        ads.add(ad);
                        latch.countDown();
                    }

                    @Override
                    public void onFailure(NendAdFullBoardLoader.FullBoardAdError error) {
                        latch.countDown();
                    }
                });
            }
            try {
                latch.await();
            } catch (InterruptedException ignore) {

            }
            return ads;
        }

        @Override
        public void deliverResult(List<NendAdFullBoard> data) {
            mAds = data;
            if (isStarted()) {
                super.deliverResult(data);
            }
        }

        @Override
        protected void onStartLoading() {
            if (mAds != null) {
                deliverResult(mAds);
            } else {
                forceLoad();
            }
        }

        @Override
        protected void onStopLoading() {
            cancelLoad();
        }

        @Override
        protected void onReset() {
            super.onReset();
            onStopLoading();
            mAds = null;
        }
    }

    private static class ContentPage implements Page {

        @Override
        public Fragment getFragment() {
            return new ContentFragment();
        }
    }

    private static class AdPage implements Page {

        private final NendAdFullBoard mAd;

        private AdPage(NendAdFullBoard ad) {
            mAd = ad;
        }

        @Override
        public Fragment getFragment() {
            AdFragment f = new AdFragment();
            f.setAd(mAd);
            return f;
        }
    }

    private static class Adapter extends FragmentPagerAdapter {

        private final List<Page> mPages;

        Adapter(FragmentManager fm, List<Page> pages) {
            super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
            mPages = pages;
        }

        @Override
        public int getItemPosition(@NonNull Object object) {
            if (object instanceof AdFragment) {
                return POSITION_NONE;
            } else {
                return super.getItemPosition(object);
            }
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            return mPages.get(position).getFragment();
        }

        @Override
        public int getCount() {
            return mPages.size();
        }
    }
}

File: Sample/kotlin/src/main/java/net/nend/sample/kotlin/fullboard/FullBoardPagerActivity.kt
package net.nend.sample.kotlin.fullboard

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.loader.app.LoaderManager
import androidx.loader.content.Loader
import androidx.viewpager.widget.PagerAdapter
import net.nend.android.NendAdFullBoard
import net.nend.android.NendAdFullBoardLoader
import net.nend.android.NendAdFullBoardView
import net.nend.sample.kotlin.databinding.ActivityFullBoardPagerBinding
import net.nend.sample.kotlin.fullboard.FullBoardMenuActivity.Companion.FULLBOARD_API_KEY
import net.nend.sample.kotlin.fullboard.FullBoardMenuActivity.Companion.FULLBOARD_LOG_TAG
import net.nend.sample.kotlin.fullboard.FullBoardMenuActivity.Companion.FULLBOARD_SPOT_ID
import java.util.concurrent.CountDownLatch

class FullBoardPagerActivity : AppCompatActivity(),
        LoaderManager.LoaderCallbacks<List<NendAdFullBoard>> {

    private lateinit var binding: ActivityFullBoardPagerBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityFullBoardPagerBinding.inflate(layoutInflater)
        setContentView(binding.root)

        LoaderManager.getInstance(this).initLoader(0, Bundle(), this)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        Handler(Looper.getMainLooper()).post {
            // Update the ad orientation.
            binding.pager.adapter?.notifyDataSetChanged()
        }
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<List<NendAdFullBoard>> {
        return AdLoader(applicationContext)
    }

    override fun onLoadFinished(loader: Loader<List<NendAdFullBoard>>, data: List<NendAdFullBoard>) {

        val pages = (0..4).map { ContentPage() }.toMutableList<Page>()
        if (AD_COUNT == data.size) {
            pages.add(2, AdPage(data[0]))
            pages.add(4, AdPage(data[1]))
        } else {
            Log.d(FULLBOARD_LOG_TAG, "Couldn't obtain two ads.")
        }
        binding.pager.adapter = Adapter(supportFragmentManager, pages)
    }

    override fun onLoaderReset(loader: Loader<List<NendAdFullBoard>>) {}

    internal interface Page {
        val fragment: Fragment
    }

    class ContentFragment : Fragment() {
        @SuppressLint("SetTextI18n")
        override fun onCreateView(
                inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
            return TextView(activity).apply {
                text = "Content"
                gravity = Gravity.CENTER
            }
        }
    }

    class AdFragment : Fragment(), NendAdFullBoardView.FullBoardAdClickListener {
        private var fullBoardAd: NendAdFullBoard? = null

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            retainInstance = true
        }

        override fun onCreateView(
                inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return if (fullBoardAd != null) {
                NendAdFullBoardView.Builder.with(activity, fullBoardAd)
                        .adClickListener(this)
                        .build()
            } else {
                null
            }
        }

        internal fun setAd(ad: NendAdFullBoard) {
            fullBoardAd = ad
        }

        override fun onClickAd(ad: NendAdFullBoard) {
            Log.d(FULLBOARD_LOG_TAG, "ClickAd")
        }
    }

    private class AdLoader(context: Context) :
            androidx.loader.content.AsyncTaskLoader<List<NendAdFullBoard>>(context) {

        private var ads: List<NendAdFullBoard>? = null
        private val loader: NendAdFullBoardLoader = NendAdFullBoardLoader(context,
                FULLBOARD_SPOT_ID, FULLBOARD_API_KEY)

        override fun loadInBackground(): List<NendAdFullBoard> {
            val latch = CountDownLatch(AD_COUNT)
            val ads = mutableListOf<NendAdFullBoard>()

            for (i in 0 until AD_COUNT) {
                loader.loadAd(object : NendAdFullBoardLoader.Callback {
                    override fun onSuccess(ad: NendAdFullBoard) {
                        ads.add(ad)
                        latch.countDown()
                    }

                    override fun onFailure(error: NendAdFullBoardLoader.FullBoardAdError) {
                        latch.countDown()
                    }
                })
            }
            try {
                latch.await()
            } catch (ignore: InterruptedException) {

            }

            return ads
        }

        override fun deliverResult(data: List<NendAdFullBoard>?) {
            ads = data
            if (isStarted) {
                super.deliverResult(data)
            }
        }

        override fun onStartLoading() {
            ads?.let {
                deliverResult(it)
            } ?: run {
                forceLoad()
            }
        }

        override fun onStopLoading() {
            cancelLoad()
        }

        override fun onReset() {
            super.onReset()
            onStopLoading()
            ads = null
        }
    }

    private inner class ContentPage : Page {

        override val fragment: Fragment
            get() = ContentFragment()
    }

    private inner class AdPage(private val ad: NendAdFullBoard) : Page {

        override val fragment: Fragment
            get() = AdFragment().apply { setAd(ad) }
    }

    private inner class Adapter(
            fm: FragmentManager, private val pages: List<Page>) : FragmentPagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

        override fun getItemPosition(`object`: Any): Int {
            return if (`object` is AdFragment) {
                PagerAdapter.POSITION_NONE
            } else {
                super.getItemPosition(`object`)
            }
        }

        override fun getItem(position: Int) = pages[position].fragment

        override fun getCount() = pages.size
    }

    companion object {
        private const val AD_COUNT = 2
    }
}